File: versions/Winlsland-1.2.3beta/project/tests/media-fixture.cpp
// Native integration producer. This is not linked into or required by the application.
#include "../src/core.h"
#include <winrt/Windows.Foundation.h>
#include <winrt/Windows.Media.h>
#include <winrt/Windows.Media.Core.h>
#include <winrt/Windows.Media.Playback.h>
using namespace wi;
using namespace winrt;
using namespace winrt::Windows::Media;
int WINAPI wWinMain(HINSTANCE,HINSTANCE,PWSTR,int){
    int argc;auto argv=CommandLineToArgvW(GetCommandLineW(),&argc);if(argc!=2){LocalFree(argv);return 1;}
    fs::path dir=argv[1];LocalFree(argv);fs::create_directories(dir);init_apartment();SetCurrentProcessExplicitAppUserModelID(L"WinIsland.MediaFixture");
    try{
        auto path=dir/L"fixture.wav";std::ofstream wave(path,std::ios::binary);
        auto word=[&](auto value){wave.write((char*)&value,sizeof(value));};
        int length=8000*2*180;wave.write("RIFF",4);word(length+36);wave.write("WAVEfmt ",8);word(16);word((short)1);word((short)1);word(8000);word(16000);word((short)2);word((short)16);wave.write("data",4);word(length);
        for(int i=0;i<length/2;i++){double level=i%32000<16000?.05:.004;word((short)(32767*level*std::sin(2*3.141592653589793*220*i/8000)));}wave.close();
        Playback::MediaPlayer player;player.Volume(0);player.CommandManager().IsEnabled(false);
        player.Source(Core::MediaSource::CreateFromUri(winrt::Windows::Foundation::Uri(path.wstring())));
        auto smtc=player.SystemMediaTransportControls();smtc.IsEnabled(true);smtc.IsPlayEnabled(true);smtc.IsPauseEnabled(true);smtc.IsNextEnabled(true);smtc.IsPreviousEnabled(true);
        bool playing=true,stopped=false,ignore=false;int track=1;double position=15,stamp=now();std::mutex mutex;std::deque<std::wstring> pending;
        auto queue=[&](std::wstring s){std::lock_guard l(mutex);pending.push_back(s);};
        auto button=smtc.ButtonPressed([&](auto&&,auto&&e){auto b=e.Button();if(b==SystemMediaTransportControlsButton::Play)queue(L"play");if(b==SystemMediaTransportControlsButton::Pause)queue(L"pause");if(b==SystemMediaTransportControlsButton::Next)queue(L"next");if(b==SystemMediaTransportControlsButton::Previous)queue(L"previous");});
        auto repeat=smtc.AutoRepeatModeChangeRequested([&](auto&&,auto&&e){queue(L"repeat="+std::to_wstring((int)e.RequestedAutoRepeatMode()));});
        auto shuffle=smtc.ShuffleEnabledChangeRequested([&](auto&&,auto&&e){queue(e.RequestedShuffleEnabled()?L"shuffle=1":L"shuffle=0");});
        auto update=[&]{auto u=smtc.DisplayUpdater();u.Type(MediaPlaybackType::Music);u.MusicProperties().Title(L"WinIsland 合成媒体 "+std::to_wstring(track));u.MusicProperties().Artist(L"测试歌手");u.MusicProperties().AlbumTitle(L"本机验证");u.Update();smtc.PlaybackStatus(playing?MediaPlaybackStatus::Playing:stopped?MediaPlaybackStatus::Stopped:MediaPlaybackStatus::Paused);SystemMediaTransportControlsTimelineProperties t;t.StartTime(std::chrono::seconds(0));t.EndTime(std::chrono::seconds(180));t.MinSeekTime(std::chrono::seconds(0));t.MaxSeekTime(std::chrono::seconds(180));t.Position(winrt::Windows::Foundation::TimeSpan((long long)(position*1e7)));smtc.UpdateTimelineProperties(t);};
        player.Play();update();writeAtomic(dir/L"media-ready.txt",std::to_string(GetCurrentProcessId()));double lifetime=now();bool exit=false;
        while(!exit&&now()-lifetime<600){auto command=dir/L"media-command.txt";if(fs::exists(command)){queue(wide(readFile(command)));fs::remove(command);}std::deque<std::wstring> actions;{std::lock_guard l(mutex);actions.swap(pending);}for(auto &c:actions){
            if(c==L"exit"){exit=true;break;}if(c==L"audio-on"){player.Volume(.12);player.IsMuted(false);continue;}if(c==L"mute"){player.IsMuted(true);continue;}if(c==L"unmute"){player.IsMuted(false);continue;}if(c==L"mode-on"){smtc.AutoRepeatMode(MediaPlaybackAutoRepeatMode::None);smtc.ShuffleEnabled(false);continue;}if(c==L"ignore-next"){ignore=true;continue;}if(c==L"disable-previous"){smtc.IsPreviousEnabled(false);continue;}
            if(c.rfind(L"repeat=",0)==0){smtc.AutoRepeatMode((MediaPlaybackAutoRepeatMode)std::stoi(c.substr(7)));continue;}if(c.rfind(L"shuffle=",0)==0){smtc.ShuffleEnabled(c==L"shuffle=1");continue;}if(c==L"next"&&ignore)continue;
            if(c==L"pause"||c==L"stop"){if(playing)position+=now()-stamp;playing=false;stopped=c==L"stop";player.Pause();}
            if(c==L"play"){stamp=now();playing=true;stopped=false;player.Play();}
            if(c==L"next"||c==L"previous"){track+=c==L"next"?1:-1;position=0;stamp=now();}
            if(c.rfind(L"seek=",0)==0){position=std::stod(c.substr(5));stamp=now();player.PlaybackSession().Position(winrt::Windows::Foundation::TimeSpan((long long)(position*1e7)));}update();
        }
        writeAtomic(dir/L"media-state.txt",std::string("Playing=")+(playing?"True":"False")+"\nTrack="+std::to_string(track)+"\nRepeat="+std::to_string((int)smtc.AutoRepeatMode())+"\nShuffle="+(smtc.ShuffleEnabled()?"True":"False")+"\nMuted="+(player.IsMuted()?"True":"False")+"\n");std::this_thread::sleep_for(std::chrono::milliseconds(100));}
        smtc.ButtonPressed(button);smtc.AutoRepeatModeChangeRequested(repeat);smtc.ShuffleEnabledChangeRequested(shuffle);smtc.IsEnabled(false);player.Close();return 0;
    }catch(const winrt::hresult_error&e){writeAtomic(dir/L"fixture-error.txt",utf8(e.message().c_str()));return 1;}
}

File: versions/Winlsland-1.2.2beta/project/MediaFixture.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Text;
using Windows.Media;
using Windows.Media.Core;
using Windows.Media.Playback;
using Forms = System.Windows.Forms;

namespace WinIsland
{
    // A silent local Windows media session used only by the explicit diagnostic mode.
    internal static class MediaFixture
    {
        internal static int Run(string directory)
        {
            Directory.CreateDirectory(directory);
            Native.SetCurrentProcessExplicitAppUserModelID("WinIsland.MediaFixture");
            string audio = Path.Combine(directory, "silent-fixture.wav");
            int length = 8000 * 2 * 180;
            using (var writer = new BinaryWriter(File.Create(audio)))
            {
                writer.Write(Encoding.ASCII.GetBytes("RIFF")); writer.Write(length + 36);
                writer.Write(Encoding.ASCII.GetBytes("WAVEfmt ")); writer.Write(16);
                writer.Write((short)1); writer.Write((short)1); writer.Write(8000); writer.Write(16000);
                writer.Write((short)2); writer.Write((short)16); writer.Write(Encoding.ASCII.GetBytes("data")); writer.Write(length);
                // Diagnostic PCM only. Muted by default; explicit audio-on tests the real
                // Windows session meter with alternating low-level tones, never UI animation.
                for (int sample = 0; sample < length / 2; sample++)
                {
                    double amplitude = sample % 32000 < 16000 ? .05 : .004;
                    writer.Write((short)(32767 * amplitude * Math.Sin(2 * Math.PI * 220 * sample / 8000)));
                }
            }
            using (var player = new MediaPlayer())
            using (var timer = new Forms.Timer { Interval = 200 })
            {
                player.Volume = 0; player.CommandManager.IsEnabled = false;
                player.Source = MediaSource.CreateFromUri(new Uri(audio));
                var controls = player.SystemMediaTransportControls;
                controls.IsEnabled = true; controls.IsPlayEnabled = true; controls.IsPauseEnabled = true;
                controls.IsNextEnabled = true; controls.IsPreviousEnabled = true;
                bool playing = true, ignoreNext = false, stopped = false; int track = 1;
                bool? requestedShuffle = null; MediaPlaybackAutoRepeatMode? requestedRepeat = null;
                controls.AutoRepeatModeChangeRequested += (s, e) => requestedRepeat = e.RequestedAutoRepeatMode;
                controls.ShuffleEnabledChangeRequested += (s, e) => requestedShuffle = e.RequestedShuffleEnabled;
                var lifetime = Stopwatch.StartNew(); var elapsed = Stopwatch.StartNew();
                double position = 15;
                Action update = delegate {
                    controls.DisplayUpdater.Type = MediaPlaybackType.Music;
                    controls.DisplayUpdater.MusicProperties.Title = "WinIsland 合成媒体 " + track;
                    controls.DisplayUpdater.MusicProperties.Artist = "测试歌手";
                    controls.DisplayUpdater.MusicProperties.AlbumTitle = "本机验证";
                    controls.DisplayUpdater.Update();
                    controls.PlaybackStatus = playing ? MediaPlaybackStatus.Playing : stopped ? MediaPlaybackStatus.Stopped : MediaPlaybackStatus.Paused;
                    controls.UpdateTimelineProperties(new SystemMediaTransportControlsTimelineProperties {
                        StartTime = TimeSpan.Zero, EndTime = TimeSpan.FromSeconds(180), MinSeekTime = TimeSpan.Zero,
                        MaxSeekTime = TimeSpan.FromSeconds(180), Position = TimeSpan.FromSeconds(position)
                    });
                };
                Action<string> command = text => {
                    if (text == "audio-on") { player.Volume = .12; player.IsMuted = false; return; }
                    if (text == "mute") { player.IsMuted = true; return; }
                    if (text == "unmute") { player.IsMuted = false; return; }
                    if (text == "mode-on") { controls.AutoRepeatMode = MediaPlaybackAutoRepeatMode.None; controls.ShuffleEnabled = false; return; }
                    if (text == "ignore-next") { ignoreNext = true; return; }
                    if (text == "disable-previous") { controls.IsPreviousEnabled = false; return; }
                    if (text == "next" && ignoreNext) return;
                    if (text == "pause" || text == "stop") { position += elapsed.Elapsed.TotalSeconds; elapsed.Reset(); playing = false; stopped = text == "stop"; player.Pause(); }
                    if (text == "play") { elapsed.Restart(); playing = true; stopped = false; player.Play(); }
                    if (text.StartsWith("seek=")) { position = double.Parse(text.Substring(5), System.Globalization.CultureInfo.InvariantCulture); player.PlaybackSession.Position = TimeSpan.FromSeconds(position); if (playing) elapsed.Restart(); }
                    if (text == "next" || text == "previous") { track += text == "next" ? 1 : -1; position = 0; elapsed.Restart(); }
                    update();
                };
                string pendingButton = null;
                controls.ButtonPressed += (s, e) => {
                    if (e.Button == SystemMediaTransportControlsButton.Play) pendingButton = "play";
                    if (e.Button == SystemMediaTransportControlsButton.Pause) pendingButton = "pause";
                    if (e.Button == SystemMediaTransportControlsButton.Next) pendingButton = "next";
                    if (e.Button == SystemMediaTransportControlsButton.Previous) pendingButton = "previous";
                };
                player.Play(); update();
                timer.Tick += delegate {
                    string path = Path.Combine(directory, "media-command.txt");
                    if (File.Exists(path))
                    {
                        string value = File.ReadAllText(path).Trim(); File.Delete(path);
                        if (value == "exit") { Forms.Application.ExitThread(); return; }
                        command(value);
                    }
                    if (pendingButton != null) { string value = pendingButton; pendingButton = null; command(value); }
                    if (requestedRepeat.HasValue) { controls.AutoRepeatMode = requestedRepeat.Value; requestedRepeat = null; }
                    if (requestedShuffle.HasValue) { controls.ShuffleEnabled = requestedShuffle.Value; requestedShuffle = null; }
                    File.WriteAllLines(Path.Combine(directory, "media-state.txt"), new[] { "Playing=" + playing, "Track=" + track,
                        "Repeat=" + controls.AutoRepeatMode, "Shuffle=" + controls.ShuffleEnabled, "Muted=" + player.IsMuted, "Volume=" + player.Volume });
                    if (lifetime.Elapsed.TotalSeconds > 180) Forms.Application.ExitThread();
                };
                timer.Start();
                File.WriteAllText(Path.Combine(directory, "media-ready.txt"), Process.GetCurrentProcess().Id.ToString());
                Forms.Application.Run();
                controls.IsEnabled = false;
            }
            return 0;
        }
    }
}
